invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java

```diff
@@ -572,90 +607,154 @@ private InvoiceWithFutureNotifications processAccountWithLockAndInputTargetDate(
-                final Set<UUID> uniqueInvoiceIds = getUniqueInvoiceIds(invoice);
-                final boolean isRealInvoiceWithItems = uniqueInvoiceIds.remove(invoice.getId());
-                final Set<UUID> adjustedUniqueOtherInvoiceId = uniqueInvoiceIds;
-
-                logInvoiceWithItems(account, invoice, actualTargetDate, adjustedUniqueOtherInvoiceId, isRealInvoiceWithItems);
+            // For non dry-run scenarii, the CBA will be recomputed in the DAO layer (https://github.com/killbill/killbill/issues/637)
+            if (isDryRun) {
+                for (final DefaultInvoice generatedInvoice : generatedInvoices) {
+                    final InvoiceItem cbaItemPostInvoicePlugins = computeCBAOnExistingInvoice(generatedInvoice, internalCallContext);
+                    if (cbaItemPostInvoicePlugins != null) {
+                        generatedInvoice.addInvoiceItem(cbaItemPostInvoicePlugins);
+                    }
+                }
+                success = true;
+            } else {
+                final List<InvoiceModelDao> invoiceModelDaos = new LinkedList<InvoiceModelDao>();
+                for (final DefaultInvoice generatedInvoice : generatedInvoices) {
+                    if (generatedInvoice.getInvoiceItems().isEmpty()) {
+                        continue;
+                    }
+                    logInvoiceWithItems(account, generatedInvoice, actualTargetDate);
 
-                // Transformation to Invoice -> InvoiceModelDao
-                final InvoiceModelDao invoiceModelDao = new InvoiceModelDao(invoice);
-                final List<InvoiceItemModelDao> invoiceItemModelDaos = transformToInvoiceModelDao(invoice.getInvoiceItems());
-                invoiceModelDao.addInvoiceItems(invoiceItemModelDaos);
+                    // Transformation to Invoice -> InvoiceModelDao
+                    final InvoiceModelDao invoiceModelDao = new InvoiceModelDao(generatedInvoice);
+                    final List<InvoiceItemModelDao> invoiceItemModelDaos = transformToInvoiceModelDao(generatedInvoice.getInvoiceItems());
+                    invoiceModelDao.addInvoiceItems(invoiceItemModelDaos);
+                    invoiceModelDaos.add(invoiceModelDao);
+                }
 
+                if (invoiceModelDaos.isEmpty()) {
+                    setFutureNotifications(account, futureAccountNotifications, internalCallContext);
+                } else {
+                    // Note: we're demultiplexing the tracking ids here (same tracking ids for all invoices generated), as we don't
+                    // really know how the plugin(s) did the grouping. If multiple invoice items were generated for a a given subscription
+                    // for instance (e.g. catch-up scenario), they might now end up on different invoices.
+                    final Set<InvoiceTrackingModelDao> trackingIds = new HashSet<InvoiceTrackingModelDao>();
```

Comment:
That part bothers me a little bit though... 

At the very least, I think we will need to change the `TrackingRecordId#equals` and `TrackingRecordId#hashCode` to not take into account the `invoiceId` (so essentially we could reuse the already existing `TrackingRecordId#isSimilarRecord`) so as to not have duplicate in our set [here](https://github.com/killbill/killbill/blob/killbill-0.22.9/invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalUsageInArrear.java#L499)

Reply:
Tracking ids are unique per subscription and record date. One thing I could do is add some amount of logic to parse the invoice items, find which tracking date falls within which service period, and dispatch these ids manually?